File: ts/src/binary_search.ts
/**
 * Performs binary search to find the index of target in a sorted array.
 * @param arr - A sorted array of integers
 * @param target - The value to search for
 * @returns The index of target if found, -1 otherwise
 *
 * @example
 * const arr = [1, 2, 3, 4, 5];
 * binarySearch(arr, 3); // returns 2
 * binarySearch(arr, 6); // returns -1
 *
 * @complexity
 * Time: O(log n)
 * Space: O(1)
 */
export function binarySearch(arr: number[], target: number): number {
    if (!arr || arr.length === 0) {
        return -1;
    }

    let left: number = 0;
    let right: number = arr.length - 1;

    while (left <= right) {
        const mid: number = Math.floor((left + right) / 2);

        if (arr[mid] === target) {
            return mid;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return -1;
}

// Example usage
const testCases: Array<{ arr: number[]; target: number }> = [
    { arr: [1, 2, 3, 4, 5], target: 3 },      // Target in middle
    { arr: [1, 2, 3, 4, 5], target: 1 },      // Target at start
    { arr: [1, 2, 3, 4, 5], target: 5 },      // Target at end
    { arr: [1, 2, 3, 4, 5], target: 6 },      // Target not present
    { arr: [], target: 1 },                    // Empty array
    { arr: [1], target: 1 },                   // Single element
];

testCases.forEach(({ arr, target }) => {
    const result = binarySearch(arr, target);
    console.log(`Array: [${arr}], Target: ${target}, Result: ${result}`);
});

File: haskell/src/BinarySearch.hs
{-|
Module      : BinarySearch
Description : Binary search implementation in Haskell
Copyright   : (c) 2024
License     : MIT
Maintainer  : example@example.com

This module provides a binary search implementation for sorted lists.
-}

module BinarySearch
    ( binarySearch
    ) where

-- | Performs binary search to find the index of target in a sorted list.
-- Returns Nothing if the target is not found.
--
-- >>> binarySearch [1,2,3,4,5] 3
-- Just 2
--
-- >>> binarySearch [1,2,3,4,5] 6
-- Nothing
--
-- >>> binarySearch [] 1
-- Nothing
--
-- >>> binarySearch [1] 1
-- Just 0
--
-- Complexity:
--   * Time: O(log n)
--   * Space: O(1)
binarySearch :: (Ord a) => [a] -> a -> Maybe Int
binarySearch [] _ = Nothing
binarySearch xs target = go xs target 0 (length xs - 1)
  where
    go :: (Ord a) => [a] -> a -> Int -> Int -> Maybe Int
    go xs target left right
      | left > right = Nothing
      | otherwise = let mid = (left + right) `div` 2
                        midVal = xs !! mid
                    in case compare midVal target of
                         EQ -> Just mid
                         LT -> go xs target (mid + 1) right
                         GT -> if mid == 0 then Nothing
                               else go xs target left (mid - 1)

-- Example usage
main :: IO ()
main = do
    -- Test cases
    let testCases = [
            ([1,2,3,4,5], 3),    -- Target in middle
            ([1,2,3,4,5], 1),    -- Target at start
            ([1,2,3,4,5], 5),    -- Target at end
            ([1,2,3,4,5], 6),    -- Target not present
            ([], 1),             -- Empty list
            ([1], 1)             -- Single element
          ]

    -- Run tests
    mapM_ (\(xs, target) -> do
        putStrLn $ "Array: " ++ show xs ++ ", Target: " ++ show target ++
                   ", Result: " ++ show (binarySearch xs target))
        ) testCases
